nativelib/src/main/scala/scala/scalanative/runtime/Backtrace.scala

```diff
@@ -15,86 +19,60 @@ import scala.annotation.tailrec
 
 import java.io.File
 import java.util.concurrent.ConcurrentHashMap
-import scala.scalanative.meta.LinktimeInfo.isMultithreadingEnabled
 import java.util.HashMap
 import java.util.AbstractMap
 
 private[runtime] object Backtrace {
-  private sealed trait Format
-  private case object MACHO extends Format
-  private case object ELF extends Format
   private case class DwarfInfo(
-      subprograms: IndexedSeq[SubprogramDIE],
+      subprograms: scala.Array[DWARF.SubprogramDIE],
       strings: DWARF.Strings,
       /** ASLR offset (minus __PAGEZERO size for macho) */
-      offset: Long,
-      format: Format
-  )
-
-  private case class SubprogramDIE(
-      lowPC: Long,
-      highPC: Long,
-      line: Int,
-      filenameAt: Option[UInt]
+      offset: Long
   )
 
-  private val MACHO_MAGIC = "cffaedfe"
-  private val ELF_MAGIC = "7f454c46"
+  private val dwarfInfo: Option[DwarfInfo] = processFile()
 
-  private val cache: AbstractMap[String, Option[DwarfInfo]] =
-    if (isMultithreadingEnabled) new ConcurrentHashMap
-    else new HashMap
-  case class Position(filename: String, line: Int)
+  case class Position(linkageName: CString, filename: String, line: Int)
   object Position {
-    final val empty = Position(null, 0)
+    final val empty = Position(null, null, 0)
   }
 
   def decodePosition(pc: Long): Position = {
-    cache.get(filename) match {
-      case None =>
-        Position.empty // cached, there's no debug section
+    dwarfInfo match {
       case Some(info) =>
-        impl(pc, info)
-      case null =>
-        processFile(filename, None) match {
-          case None =>
-            // there's no debug section, cache it so we don't parse the exec file any longer
-            cache.put(filename, None)
-            Position.empty
-          case file @ Some(info) =>
-            cache.put(filename, file)
-            impl(pc, info)
+        // The address (DW_AT_(low|high)_address) in debug information has the file offset (the offset in the executable + __PAGEZERO in macho).
+        // While the pc address retrieved from libunwind at runtime has the location of the memory into the virtual memory
+        // at runtime. which has a random offset (called ASLR offset or slide) that is different for every run because of
+        // Address Space Layout Randomization (ASLR) when the executable is built as PIE.
+        // Subtract the offset to match the pc address from libunwind (runtime) and address in debug info (compile/link time).
+        val address = pc - info.offset
+        val position = for {
+          subprogram <- search(info.subprograms, address)
+          filenameAt <- subprogram.filenameAt
+          linkageNameAt <- subprogram.linkageNameAt
+        } yield {
+          val filename = info.strings.read(filenameAt)
+          val linkageName =
+            info.strings.buf.asInstanceOf[ByteArray].at(linkageNameAt.toInt)
```

Comment:
Not very important but with `import scala.scalanative.unsafe.UnsafeRichArray` we can use extension instead of explicit cast.  If we're sure `linkageNameAt.toInt` would always be in valid bounds, we can even use `atUnsafe` to skip redundant bound check. 

```suggestion
            info.strings.buf.at(linkageNameAt.toInt)
```